### PR TITLE
feat(conductor): add da sync

### LINKED ISSUE #470: Conductor: Add sync functionality to `data_availability::Reader`

### LINKED ISSUE #473: Conductor: add initial da block height env var
Similar to #445, a new env var should be added to control the starting sync block height when syncing from DA

### LINKED ISSUE #547: conductor: refactor start/stop sync behavior for DA reader when using `FirmOnly`
When merging the v1alpha2 api changes for conductor, we found a bug where the DA reader was never starting because the sync was only ended in sequencer reader, which was never being created (thus sync never completed) when running Conductor in `FrimOnly` mode.

This code here: https://github.com/astriaorg/astria/blob/4cb88f7aa69358805fc5d70938434a81ff210435/crates/astria-conductor/src/conductor.rs#L114
was the fix. 

This issue is for refactoring this fix to more correctly and understandably chose the correct sync settings and startup procedure based on the execution commit level setting.

### PR DESCRIPTION
**Pull request by sambukowski**
## Summary
Conductor needs to be able to sync from DA and from Sequencer. This pr adds the ability to sync from DA.

## Changes
- [x] add da initial block height to config
- [x] add initial da block height to da::Reader startup
- [x] add a find_da_start_block fn using the firm commit from the rollup
- [x] add a get_sequencer_data_from_da fn that uses a future stream to pull and verify 
- [x] reorg data_availability files
- [x] change the da::Reader run until stopped to have a sync aspect
- [x] add sync_done to the da reader 
- [x] update Conductor run_until_stopped to properly set useful variables for initializing and starting the da and seq Readers  
- [x] Confirm actor startup within conductor is correct when running at different execution commit levels
- [x] add `daBlockRange` to env vars to control search area on da when starting da sync

## Testing
Unit tests still passing
Manual testing using dev-cluster

## Breaking Changelist
- New env vars added to config. Will require chart updates

## Related Issues

closes #470 
closes #473 
closes #547

### CODE REVIEW

File: crates/astria-conductor/local.env.example [hunk 1]
```diff
@@ -32,3 +32,6 @@ ASTRIA_CONDUCTOR_SEQUENCER_URL="ws://127.0.0.1:26657/websocket"
 
 # the sequencer block height that the rollup's first block was in
 ASTRIA_CONDUCTOR_INITIAL_SEQUENCER_BLOCK_HEIGHT=1
+
+# the da block height that the rollup stated at
```
Comment (noot) [inline crates/astria-conductor/local.env.example:36]:
```suggestion
# the data availability block height that the rollup's first block was in
```

File: crates/astria-conductor/src/conductor.rs [hunk 2]
```diff
@@ -70,10 +71,20 @@ pub struct Conductor {
...
 
     /// The channel over which the sequencer reader task notifies conductor that sync is completed.
-    sync_done: Fuse<oneshot::Receiver<()>>,
+    seq_sync_done: Fuse<oneshot::Receiver<()>>,
+
+    /// The channel over which the sequencer reader task notifies conductor that sync is completed.
+    da_sync_done: Fuse<oneshot::Receiver<()>>,
+
+    /// The data availability reader that is spawned after sync is completed.
+    /// Constructed if constructed if `disable_finalization = false`.
```
Comment (noot) [inline crates/astria-conductor/src/conductor.rs:80]:
```suggestion
    /// Constructed if `disable_finalization = false`.
```

File: crates/astria-conductor/src/conductor.rs [hunk 3]
```diff
@@ -96,26 +109,32 @@ impl Conductor {
...
+        let (executor_tx, soft_commit_height, firm_commit_height) = {
             let (block_tx, block_rx) = mpsc::unbounded_channel();
             let (shutdown_tx, shutdown_rx) = oneshot::channel();
             let executor = Executor::new(
                 &cfg.execution_rpc_url,
                 ChainId::new(cfg.chain_id.as_bytes().to_vec())
                     .wrap_err("failed to create chain ID")?,
-                // the sequencer block that contains the first rollup block
-                cfg.initial_sequencer_block_height,
+                cfg.initial_sequencer_block_height, // the sequencer block the rollup was start on
```
Comment (noot) [inline crates/astria-conductor/src/conductor.rs:119]:
```suggestion
                cfg.initial_sequencer_block_height, // the sequencer block that contains the first rollup block
```

File: crates/astria-conductor/src/conductor.rs [hunk 4]
```diff
@@ -124,14 +143,33 @@ impl Conductor {
...
+                info!("only syncing from sequencer");
+                // kill the DA sync to only execute from sequencer
+                let (sync_done_tx, sync_done_rx) = oneshot::channel();
+                da_sync_done = sync_done_rx.fuse();
+                let _ = sync_done_tx.send(());
+            }
+            CommitLevel::SoftAndFirm => {
+                info!("syncing from DA then sequencer");
+                // when running in soft and firm mode, a sync cycle from both DA
+                // and sequencer are used. No need to kill any of the syncs
```
Comment (noot) [inline crates/astria-conductor/src/conductor.rs:164]: why do we need to sync blocks from both? isn't it sufficient to sync from block 1 -> firm from DA, then firm -> soft from sequencer?

File: crates/astria-conductor/src/conductor.rs [hunk 5]
```diff
@@ -183,8 +224,9 @@ impl Conductor {
                 bearer_token: Some(cfg.celestia_bearer_token),
                 poll_interval: std::time::Duration::from_secs(3),
             };
-            // TODO ghi(https://github.com/astriaorg/astria/issues/470): add sync functionality to data availability reader
-            let reader = data_availability::Reader::new(
+            let da_reader = data_availability::Reader::new(
+                cfg.initial_da_block_height,
+                firm_commit_height,
```
Comment (noot) [inline crates/astria-conductor/src/conductor.rs:229]: why are both of these needed? shouldn't only `firm_commit_height` be needed?

File: crates/astria-conductor/src/conductor.rs [hunk 6]
```diff
@@ -141,24 +179,27 @@ impl Conductor {
...
-                sync_start_block_height,
+            let seq_reader = sequencer::Reader::new(
+                soft_commit_height,
                 sequencer_client_pool.clone(),
                 shutdown_rx,
                 executor_tx.clone(),
                 sync_done_tx,
             );
-            tasks.spawn(Self::SEQUENCER, sequencer_reader.run_until_stopped());
+            sequencer_reader = Some(seq_reader);
```
Comment (noot) [inline crates/astria-conductor/src/conductor.rs:189]: why not put this in `tasks.spawn` as before?

File: crates/astria-conductor/src/conductor.rs [hunk 7]
```diff
@@ -193,23 +235,34 @@ impl Conductor {
                     cfg.chain_id.as_ref(),
                 ),
                 shutdown_rx,
+                sync_done_tx,
             )
             .await
             .wrap_err("failed constructing data availability reader")?;
-            data_availability_reader = Some(reader);
+            data_availability_reader = Some(da_reader);
```
Comment (noot) [inline crates/astria-conductor/src/conductor.rs:242]: why not put this in `tasks.spawn`?

File: crates/astria-conductor/src/conductor.rs [hunk 8]
```diff
@@ -226,23 +279,42 @@ impl Conductor {
...
                     }
                 }
 
+                // Start the sequencer reader
+                res = &mut self.seq_sync_done, if self.da_sync_done.is_terminated() && !self.seq_sync_done.is_terminated() => {
+                    match res {
+                        Ok(()) => info!("received sync-complete signal from DA reader"),
+                        Err(e) => {
+                            let error = &e as &(dyn std::error::Error + 'static);
+                            warn!(error, "DA sync-complete channel failed prematurely");
```
Comment (noot) [inline crates/astria-conductor/src/conductor.rs:306]:
```suggestion
                            warn!(error, "sequencer sync-complete channel failed prematurely");
```

File: crates/astria-conductor/src/conductor.rs [hunk 9]
```diff
@@ -226,23 +279,42 @@ impl Conductor {
...
                             Self::DATA_AVAILABILITY,
                             data_availability_reader.run_until_stopped(),
                         );
                     }
                 }
 
+                // Start the sequencer reader
+                res = &mut self.seq_sync_done, if self.da_sync_done.is_terminated() && !self.seq_sync_done.is_terminated() => {
+                    match res {
+                        Ok(()) => info!("received sync-complete signal from DA reader"),
```
Comment (noot) [inline crates/astria-conductor/src/conductor.rs:303]:
```suggestion
                        Ok(()) => info!("received sync-complete signal from sequencer reader"),
```

File: crates/astria-conductor/src/config.rs [hunk 10]
```diff
@@ -46,6 +46,9 @@ pub struct Config {
     /// The Sequencer block height that the rollup genesis block was in
     pub initial_sequencer_block_height: u32,
 
+    /// The DA block height that the rollup genesis block was in
```
Comment (noot) [inline crates/astria-conductor/src/config.rs:49]:
```suggestion
    /// The DA block height that the rollup's first block was in
```

File: crates/astria-conductor/src/data_availability/mod.rs [hunk 11]
```diff
@@ -125,14 +142,20 @@ pub(crate) struct CelestiaReaderConfig {
...
+        firm_commit_height: u32,
         celestia_config: CelestiaReaderConfig,
         executor_tx: executor::Sender,
         block_verifier: BlockVerifier,
         sequencer_namespace: Namespace,
         rollup_namespace: Namespace,
         shutdown: oneshot::Receiver<()>,
+        sync_done: oneshot::Sender<()>,
     ) -> eyre::Result<Self> {
+        info!("creating da reader");
```
Comment (noot) [inline crates/astria-conductor/src/data_availability/mod.rs:157]: remove

File: crates/astria-conductor/src/conductor.rs [hunk 12]
```diff
@@ -193,23 +235,34 @@ impl Conductor {
                     cfg.chain_id.as_ref(),
                 ),
                 shutdown_rx,
+                sync_done_tx,
             )
             .await
             .wrap_err("failed constructing data availability reader")?;
-            data_availability_reader = Some(reader);
+            data_availability_reader = Some(da_reader);
+            shutdown_channels.insert(Self::DATA_AVAILABILITY, shutdown_tx);
```
Comment (noot) [inline crates/astria-conductor/src/conductor.rs:243]: the task was not spawned yet - put the shutdown channel in the map when it's spawned. same for line 190

File: crates/astria-conductor/src/data_availability/mod.rs [hunk 13]
```diff
@@ -163,18 +186,54 @@ impl Reader {
...
+            sync_done: Some(sync_done),
         })
     }
 
     #[instrument(skip(self))]
     pub(crate) async fn run_until_stopped(mut self) -> eyre::Result<()> {
         info!("Starting reader event loop.");
 
-        // TODO ghi(https://github.com/astriaorg/astria/issues/470): add sync functionality to data availability reader
+        // TODO ghi(https://github.com/astriaorg/astria/issues/470): add sync
```
Comment (noot) [inline crates/astria-conductor/src/data_availability/mod.rs:200]: is this closed by this PR?

File: crates/astria-conductor/src/data_availability/mod.rs [hunk 14]
```diff
@@ -163,18 +186,54 @@ impl Reader {
...
+            sync_start_height,
+            latest_height,
+            self.sequencer_namespace,
+            self.executor_tx.clone(),
+            self.celestia_client.clone(),
+            self.block_verifier.clone(),
+        )
+        .boxed()
+        .fuse();
+        // self.fetch_sequencer_blobs_up_to_latest_height(Ok(Ok(height)));
```
Comment (noot) [inline crates/astria-conductor/src/data_availability/mod.rs:228]:
```suggestion
```

File: crates/astria-conductor/src/data_availability/mod.rs [hunk 15]
```diff
@@ -163,18 +186,54 @@ impl Reader {
...
+            self.celestia_client.clone(),
+            self.initial_da_block_height,
+            self.firm_commit_height,
+            self.sequencer_namespace,
+        )
+        .await;
+
+        let latest_height = u32::try_from(
+            find_latest_height(self.celestia_client.clone())
+                .await?
```
Comment (noot) [inline crates/astria-conductor/src/data_availability/mod.rs:213]: wrap_err

File: crates/astria-conductor/src/conductor.rs (hunk 5 above)
Comment (sambukowski) [reply crates/astria-conductor/src/conductor.rs:229]: These are both needed because the initial da block height is the celestia block height, and the firm height is the height from the sequencer. To be able to find the da block we need to start syncing from we need the firm commit when looking inside the da blocks, and the da block height to better find which da block to look in.

File: crates/astria-conductor/src/conductor.rs (hunk 4 above)
Comment (sambukowski) [reply crates/astria-conductor/src/conductor.rs:164]: Yeah that was my intention. I'll update the comment to reflect that.

File: crates/astria-conductor/src/data_availability/mod.rs (hunk 13 above)
Comment (sambukowski) [reply crates/astria-conductor/src/data_availability/mod.rs:200]: yup! will delete

File: crates/astria-conductor/src/data_availability/mod.rs [hunk 16]
```diff
@@ -160,18 +189,116 @@ impl Reader {
             block_verifier,
             sequencer_namespace,
             rollup_namespace,
+            initial_seq_block_height: initial_seq_height,
+            initial_da_block_height: initial_da_height,
+            firm_commit_height,
+            da_block_range,
             shutdown,
+            sync_done,
```
Comment (SuperFluffy) [inline crates/astria-conductor/src/data_availability/mod.rs:197]: Why is this necessary?

File: crates/astria-conductor/src/conductor.rs [hunk 17]
```diff
@@ -45,20 +51,39 @@ use crate::{
...
 pub struct Conductor {
-    /// The data availability reader that is spawned after sync is completed.
-    /// Constructed if constructed if `disable_finalization = false`.
+    /// The sequencer reader that is spawned after either after DA sync is completed,
+    /// or when running in "SoftOnly".
+    sequencer_reader: Option<sequencer::Reader>,
+
+    /// The data availability syncer that is spawned for DA sync, or after
+    /// sequencer sync has been completed.
+    data_availability_syncer: Option<data_availability::Reader>,
```
Comment (SuperFluffy) [inline crates/astria-conductor/src/conductor.rs:76]: I am not sure I understand the design. Why is there a separate data availability syncer and reader? I see that conductor has a `sync_done` field, but it seems unnecessary given the 2 separate actors here?

File: crates/astria-conductor/src/config.rs [hunk 18]
```diff
@@ -46,6 +46,12 @@ pub struct Config {
     /// The Sequencer block height that the rollup genesis block was in
     pub initial_sequencer_block_height: u32,
 
+    /// The DA block height that the rollup's first block was in
+    pub initial_da_block_height: u32,
```
Comment (SuperFluffy) [inline crates/astria-conductor/src/config.rs:50]: Please don't abbreviate

File: crates/astria-conductor/src/data_availability/block_verifier.rs [hunk 19]
```diff
@@ -265,8 +265,8 @@ fn verify_vote_signature(
     public_key
         .verify(
             &signature,
-            &tendermint_proto::types::CanonicalVote::try_from(canonical_vote)
-                .wrap_err("failed to turn commit canonical vote into proto type")?
+            &tendermint_proto::types::CanonicalVote::from(canonical_vote)
+                // .wrap_err("failed to turn commit canonical vote into proto type")?
```
Comment (SuperFluffy) [inline crates/astria-conductor/src/data_availability/block_verifier.rs:269]: cleanup

File: crates/astria-conductor/src/data_availability/mod.rs [hunk 20]
```diff
@@ -160,18 +189,116 @@ impl Reader {
...
+            .boxed()
+            .fuse();
+        } else {
+            info!(
+                "current da block height is the same as the initial da block height; skipping da \
+                 sync"
+            );
+        }
+
+        loop {
```
Comment (SuperFluffy) [inline crates/astria-conductor/src/data_availability/mod.rs:263]: Will this loop ever end after sync is complete?

File: crates/astria-conductor/src/sequencer/mod.rs [hunk 21]
```diff
@@ -377,7 +389,10 @@ fn schedule_for_forwarding_or_resubscribe(
                     "block received from sequencer subscription was bad; dropping it"
                 );
             }
-            Ok(block) => pending_blocks.push_back(futures::future::ready(block)),
+            Ok(block) => {
+                info!(at_height = %block.header().height, "received new block from sequencer subscription");
```
Comment (SuperFluffy) [inline crates/astria-conductor/src/sequencer/mod.rs:393]: The `from sequencer subscription` is redundant because this information is derived from the environment

File: crates/astria-conductor/src/conductor.rs [hunk 22]
```diff
@@ -235,16 +376,62 @@ impl Conductor {
                     info!("reloading is currently not implemented");
                 }
 
-                res = &mut self.sync_done, if !self.sync_done.is_terminated() => {
+                // Start the sequencer reader
+                res = &mut self.da_sync_done, if !self.da_sync_done.is_terminated() && !self.execution_commit_level.is_firm_only() => {
```
Comment (SuperFluffy) [inline crates/astria-conductor/src/conductor.rs:380]: This block is very complex and I don't think I understand what is going on here.

File: crates/astria-conductor/src/data_availability/mod.rs [hunk 23]
```diff
@@ -160,18 +189,116 @@ impl Reader {
...
+        async fn ready() -> eyre::Result<()> {
+            Ok(())
+        }
+        let mut sync = ready().boxed().fuse();
+
+        info!("syncing from DA");
+
+        // check that the initial da block height is less than the current block height
+        let current_block_height = u32::try_from(self.current_block_height.value())
+            .expect("casting from u64 to u32 failed");
```
Comment (SuperFluffy) [inline crates/astria-conductor/src/data_availability/mod.rs:214]: `expect` should contain information about why this cast failed.

File: crates/astria-conductor/src/data_availability/sync.rs [hunk 24]
```diff
@@ -0,0 +1,191 @@
...
+) -> Result<Vec<CelestiaSequencerBlob>, Report> {
+    let retry_config = RetryFutureConfig::new(10)
+        .exponential_backoff(Duration::from_secs(5))
+        .max_delay(Duration::from_secs(60))
+        .on_retry(|attempt, next_delay: Option<Duration>, error: &Report| {
+            let wait_duration = next_delay
+                .map(humantime::format_duration)
+                .map(tracing::field::display);
+            let error: &(dyn std::error::Error + 'static) = error.as_ref();
+            let error = error.to_string();
```
Comment (SuperFluffy) [inline crates/astria-conductor/src/data_availability/sync.rs:85]: please don't do ths. log the error using its `tracing::field::Value` implementation

File: crates/astria-conductor/src/data_availability/sync.rs [hunk 25]
```diff
@@ -0,0 +1,191 @@
...
+                );
+            }
+        });
+
+    tryhard::retry_fn(|| async {
+        celestia_client
+            .get_sequencer_blobs(height, sequencer_namespace)
+            .await
+            .wrap_err("failed to fetch sequencer data from celestia")
+            // .map(|rsp| rsp.datas)
```
Comment (SuperFluffy) [inline crates/astria-conductor/src/data_availability/sync.rs:102]: cleanup

File: crates/astria-conductor/src/data_availability/sync.rs [hunk 26]
```diff
@@ -0,0 +1,191 @@
...
+use super::block_verifier::BlockVerifier;
+use crate::{
+    data_availability::{
+        send_sequencer_subsets,
+        verify_sequencer_blobs_and_assemble_rollups,
+    },
+    executor,
+};
+
+pub(crate) async fn find_da_block_with_sequencer_data(
```
Comment (SuperFluffy) [inline crates/astria-conductor/src/data_availability/sync.rs:36]: What does this function do? Loop from `start_height..start_height+da_block_range` until it finds the first sequencer namespace?

I think it should probably be called `find_height_with_sequencer_blob`

File: crates/astria-conductor/src/data_availability/sync.rs [hunk 27]
```diff
@@ -0,0 +1,191 @@
...
+        let sequencer_blobs = celestia_client
+            .get_sequencer_blobs(height, sequencer_namespace)
+            .await;
+
+        match sequencer_blobs {
+            Ok(_blobs) => {
+                info!(height = %height.value(), "found sequencer blob");
+                break height;
+            }
+            Err(error) => {
```
Comment (SuperFluffy) [inline crates/astria-conductor/src/data_availability/sync.rs:59]: This might have just failed due to network issues. Retry in that case?

File: crates/astria-conductor/src/data_availability/mod.rs [hunk 28]
```diff
@@ -334,36 +464,110 @@ impl Reader {
...
-            Err(e) => {
-                let error = &e as &(dyn std::error::Error + 'static);
-                warn!(error, "task processing sequencer data failed");
-                return Ok(());
+async fn find_latest_height(celestia_client: HttpClient) -> eyre::Result<Height> {
+    use celestia_client::celestia_rpc::HeaderClient;
+    Ok(celestia_client.header_network_head().await?.header.height)
+}
+
+#[instrument(skip_all, fields(height))]
```
Comment (SuperFluffy) [inline crates/astria-conductor/src/data_availability/mod.rs:474]: this probably doesn't compile right now? I don't see `height` in the function arguments

File: crates/astria-conductor/src/data_availability/mod.rs [hunk 29]
```diff
@@ -334,36 +464,110 @@ impl Reader {
...
-    fn send_sequencer_subsets(
-        &self,
-        sequencer_subsets_res: Result<eyre::Result<Vec<SequencerBlockSubset>>, JoinError>,
-    ) -> eyre::Result<()> {
-        let subsets = match sequencer_subsets_res {
-            Err(e) => {
-                let error = &e as &(dyn std::error::Error + 'static);
-                warn!(error, "task processing sequencer data failed");
-                return Ok(());
+async fn find_latest_height(celestia_client: HttpClient) -> eyre::Result<Height> {
```
Comment (SuperFluffy) [inline crates/astria-conductor/src/data_availability/mod.rs:469]: Can this function body just be inlined?

File: crates/astria-conductor/src/conductor.rs [hunk 30]
```diff
@@ -235,16 +376,62 @@ impl Conductor {
...
+                    if let Some(sequencer_reader) = self.sequencer_reader.take() {
+
+                        // FIXME: waiting here is a tempory solution. better
+                        // solution would be to not send the da sync complete
+                        // until all tasks for da sync are completed.
+                        // implementing the proper solution would require
+                        // communicating with the executor to make sure that the
+                        // execution queue is empty.
+                        info!("waiting for 10 seconds for async tasks to finish");
+                        time::sleep(Duration::from_secs(10)).await;
```
Comment (SuperFluffy) [inline crates/astria-conductor/src/conductor.rs:399]: Wouldn't it be ok to just populate the executor queue from the syncer? All new blocks should just be added to the tail?

I'd prefer to address that in this PR.

**Review by SuperFluffy — CHANGES_REQUESTED** [review]
Left a few remarks. The PR needs some cleanup in places, I see many commented out lines.

I think it might also not compile at the moment? 

Right now I have a hard time understanding following the code to see how in which order syncing is performed, and how syncing 

Can you add some text to the PR that explains the flow?

File: crates/astria-conductor/src/data_availability/sync.rs [hunk 31]
```diff
@@ -0,0 +1,191 @@
...
+pub(crate) async fn find_da_block_with_sequencer_data(
+    start_height: Height,
+    celestia_client: HttpClient,
+    sequencer_namespace: Namespace,
+    da_block_range: u32,
+) -> Height {
+    let mut height = start_height;
+    let mut loop_count = 0;
+    loop {
+        assert!(
```
[the remark this answers is not in the document]
Comment (sambukowski) [reply crates/astria-conductor/src/data_availability/sync.rs:45]: Crashing the program here is intentional. If no rollup data is seen within the search window (N blocks on da), something went wrong with posting the sequencer data to da.

[the remark this answers is not in the document]
Comment (SuperFluffy) [reply crates/astria-conductor/src/data_availability/sync.rs:45]: The only panics that are permitted in the code is if invariants are violated that the program cannot recover from.

This is not such a case. Break the loop, return an error, and report it at the top level

File: crates/astria-conductor/local.env.example [hunk 32]
```diff
@@ -33,6 +33,12 @@ ASTRIA_CONDUCTOR_SEQUENCER_URL="ws://127.0.0.1:26657/websocket"
 # the sequencer block height that the rollup's first block was in
 ASTRIA_CONDUCTOR_INITIAL_SEQUENCER_BLOCK_HEIGHT=1
 
+# the data availability block height that the rollup's first block was in
+ASTRIA_CONDUCTOR_INITIAL_DA_BLOCK_HEIGHT=1
+
+# The number of block on DA in which the first sequencer block for the rollup should be found
+ASTRIA_CONDUCTOR_DA_BLOCK_RANGE=100
```
[the remark this answers is not in the document]
Comment (sambukowski) [reply crates/astria-conductor/local.env.example:40]: I renamed this to `ASTRIA_CONDUCTOR_CELESTIA_SEARCH_WINDOW`, calling it a range was a little misleading.
The search window is the number of blocks on Celestia where the next sequencer block must appear, otherwise the chain cannot continue.
The initial block height is the block number on da where the first sequencer block for the rollup should be.

File: crates/astria-conductor/src/data_availability/sync.rs [hunk 33]
```diff
@@ -0,0 +1,121 @@
...
+        FutureExt as _,
+        StreamExt as _,
+    };
+
+    let mut height_stream = futures::stream::iter(start_sync_height..end_sync_height);
+    let mut block_stream = FuturesOrdered::new();
+
+    'sync: loop {
+        let client = client.clone();
+        let block_verifier = block_verifier.clone();
```
[the remark this answers is not in the document]
Comment (sambukowski) [reply crates/astria-conductor/src/data_availability/sync.rs:131]: I believe this is because of the `async move`s in the `select!`. 
When those let statements are removed, you get a `use of moved value` error. Should the `get_new_sequencer_block_data_with_retry` and `verify_sequencer_blobs_and_assemble_rollups` functions be changed to take the `BlockVerifyer` and the `HttpClient` as references? Or is there a specific reason not to do that?

[the remark this answers is not in the document]
Comment (SuperFluffy) [reply crates/astria-conductor/src/data_availability/sync.rs:131]: clone these objects right where you need them and then `async move` them into the blocks.


Also, the async blocks I am seeing also do yet another `clone` on the verifier and client, which is redundant

File: crates/astria-conductor/src/data_availability/mod.rs (hunk 28 above)
Comment (sambukowski) [reply crates/astria-conductor/src/data_availability/mod.rs:485]: For some reason this did compile and seems to work fine. The `fields(height)` has been removed.
